File: yandex_back_prep_python/leetcode/404. Sum of Left Leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        self.res = 0

        def dfs(root, source):
            if not root:
                return

            if source == "left" and not root.left and not root.right:
                self.res += root.val

            dfs(root.left, "left")
            dfs(root.right, "right")

        dfs(root, "")

        return self.res

File: yandex_back_prep_cpp/404. Sum of Left Leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int sumOfLeftLeaves(TreeNode *root)
    {
        this->dfs(root, ' ');
        return this->leaves_count;
    }

private:
    int leaves_count = 0;
    void dfs(TreeNode *root, char dist)
    {
        if (root == nullptr)
        {
            return;
        }

        if (dist == 'L' && root->left == nullptr && root->right == nullptr)
        {
            this->leaves_count += root->val;
            return;
        }

        dfs(root->left, 'L');
        dfs(root->right, 'R');
    }
};